plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java

```diff
@@ -68,6 +69,167 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
+    }
+
+    @Test
+    public void testTopNPushdown()
+    {
+        if (!hasBehavior(SUPPORTS_TOPN_PUSHDOWN)) {
+            assertThat(query("SELECT orderkey FROM orders ORDER BY orderkey limit 10"))
+                    .ordered()
+                    .isNotFullyPushedDown(TopNNode.class);
+            return;
+        }
+
+        Session session = topNPushdownEnabled(getSession());
+        assertThat(query(session, "SELECT orderkey FROM orders ORDER BY orderkey limit 10"))
+                .ordered()
+                .isFullyPushedDown();
+
+        assertThat(query(session, "SELECT orderkey FROM orders ORDER BY orderkey DESC limit 10"))
+                .ordered()
+                .isFullyPushedDown();
+
+        // multiple sort columns with different orders
+        assertThat(query(session, "SELECT * FROM orders ORDER BY shippriority DESC, totalprice ASC LIMIT 10"))
+                .ordered()
+                .isFullyPushedDown();
+
+        // TopN over aggregation column
+        assertThat(query(session, "SELECT sum(totalprice) AS total FROM orders GROUP BY custkey ORDER BY total DESC LIMIT 10"))
+                .ordered()
+                .isFullyPushedDown();
+
+        // TopN over TopN
+        assertThat(query(session, "SELECT orderkey, totalprice FROM (SELECT orderkey, totalprice FROM orders ORDER BY 1, 2 LIMIT 10) ORDER BY 2, 1 LIMIT 5"))
+                .ordered()
+                .isFullyPushedDown();
+
+        assertThat(query(session,
+                "SELECT orderkey, totalprice " +
+                "FROM (SELECT orderkey, totalprice FROM (SELECT orderkey, totalprice FROM orders ORDER BY 1, 2 LIMIT 10) " +
+                "ORDER BY 2, 1 LIMIT 5) ORDER BY 1, 2 LIMIT 3"))
+                .ordered()
+                .isFullyPushedDown();
+
+        // TopN over limit
+        assertThat(query(session, "SELECT orderkey, totalprice FROM (SELECT orderkey, totalprice FROM orders LIMIT 20) ORDER BY totalprice ASC LIMIT 5"))
+                .ordered()
+                .isFullyPushedDown();
+
+        // TopN over limit with filter
+        assertThat(query(session,
+                "SELECT orderkey, totalprice " +
+                "FROM (SELECT orderkey, totalprice FROM orders WHERE orderpriority = '1-URGENT' LIMIT 20) " +
+                "ORDER BY totalprice ASC LIMIT 5"))
+                .ordered()
+                .isFullyPushedDown();
+
+        // TopN over aggregation with filter
+        assertThat(query(session,
+                "SELECT * " +
+                "FROM (SELECT SUM(totalprice) as sum, custkey AS total FROM orders GROUP BY custkey HAVING COUNT(*) > 3) " +
+                "ORDER BY sum DESC LIMIT 10"))
+                .ordered()
+                .isFullyPushedDown();
+    }
+
+    @Test
+    public void testCaseSensitiveTopNPushdown()
+    {
+        if (!hasBehavior(SUPPORTS_TOPN_PUSHDOWN)) {
+            assertThat(query("SELECT orderkey FROM orders ORDER BY orderkey limit 10"))
```

Comment:
this is meaningless without session which enables top-n pushdown.

Reply:
Do you mean running this test is useless without a topN enabled session?

Or if you mean this specific assertion then sadly this assertion is only meant to test if topN is supported so that we can proceed with more in-depth testing.

Reply:
I meant this specific assertion

Reply:
It looks like this assertion does not really validate anything as it will always pass, no matter if connector supports or does not support topn pushdown (as topn pushdown is disabled here)

Reply:
Good catch @losipiuk. The test should ideally be the reverse then - assert that TopN is retained even with a topn-enabled-session?